normalizer_lib.py

```diff
@@ -1,205 +1,224 @@
-                       closure("." + GRAPHEMES.plus, 1, 2) +
-                       ("/" | ("/" + union(GRAPHEMES, NUMERALS)).star))
+                       union(self.graphemes, self.numerals).plus +
+                       closure("." + self.graphemes.plus, 1, 2) +
+                       ("/" + union(self.graphemes, self.numerals).star).star +
+                       self.final_punctuation.star).optimize()
 
         # For times. Two numbers, a colon, followed by two more numbers,
         # optionally followed by another colon and two numbers.
         # Typically e.g. 12:25, but also 12:25:04 for seconds.
         # Will allow non-standard times like 40:70:80 !
-        time = (closure(NUMERALS, 1, 2) +
-                closure(":" + closure(NUMERALS, 1, 2), 1, 2))
+        times = (closure(self.numerals, 1, 2) +
+                 closure(":" + closure(self.numerals, 1, 2), 1, 2)).optimize()
 
         # For large numbers. Allows either 1-6 numerals, e.g. 150000;
         # 1-6 numerals followed by a decimal separator and up to 4 decimals,
         # e.g. 2552.4575; or allows 1-3 numerals followed by a grouping
         # separator, 1-3 more numerals, a decimal separator, and 1-4 more
         # numerals, e.g. 120,000.65
         # Will allow non-standard things like 50,00,00 !
-        fancy_numbers = (closure(NUMERALS, 1, 6) |
-                         (closure(NUMERALS, 1, 6)
-                          + union(",", ".") + closure(NUMERALS, 1, 4)) |
-                         ((closure(NUMERALS, 1, 3) +
+        fancy_numbers = (closure(self.numerals, 1, 6) |
+                         (closure(self.numerals, 1, 6)
+                          + union(",", ".") + closure(self.numerals, 1, 4)) |
+                         ((closure(self.numerals, 1, 3) +
                            union(",", ".")).ques +
-                          closure(NUMERALS, 1, 3) +
+                          closure(self.numerals, 1, 3) +
                           (union(",", ".") +
-                           closure(NUMERALS, 1, 4).ques)))
-
-        return union(token, email_address, web_address, time, fancy_numbers)
-
-
-# Remove all extra whitespace between words
-# e.g. "hi      there" -> "hi there"
-
-REMOVE_EXTRA_WHITESPACE = cdrewrite(
-    pynutil.delete(byte.SPACE),
-    "",
-    byte.SPACE,
-    SIGMA_STAR)
+                           closure(self.numerals, 1, 4).ques))).optimize()
+        return union(token, email_address, web_address, times, fancy_numbers).optimize()
 
 
-# Language-specific formatting fixes
+    # Remove all extra whitespace between words
+    # e.g. "hi      there" -> "hi there"
 
-try:
-    LANGUAGE_SPECIFIC_NORM = LANGUAGE.LANGUAGE_SPECIFIC_PREPROCESSING
-except Exception:
-    LANGUAGE_SPECIFIC_NORM = cdrewrite(cross("", ""),
-                                       "",
-                                       "",
-                                       SIGMA_STAR)
 
+    def remove_extra_whitespace(self) -> Fst:
```

Comment:
Not sure if this need to be a method, you could also just define it in the init of the class and call it where you need it.

Reply:
Good point. I refactored things a bit and put the FSTs in the class init, so now none of those are methods.